File: beveren_health/beveren_health/customize/sales_invoice.py
# Copyright (c) 2026, Beveren Software and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import cint, date_diff, getdate, today

from beveren_health.beveren_health.customize.dispensing_lot import (
	process_sales_invoice_dispensing_lots,
	reverse_sales_invoice_dispensing_lots,
	validate_sales_invoice_dispensing_lots,
)


def validate_return_restrictions(doc, method):
	if not doc.is_return or not doc.return_against:
		return

	return_date = getdate(doc.posting_date)
	original_invoice = frappe.get_doc("Sales Invoice", doc.return_against)
	invoice_date = getdate(original_invoice.posting_date)
	days_diff = date_diff(return_date, invoice_date)

	for item in original_invoice.items:
		item_group = frappe.db.get_value("Item", item.item_code, "item_group")
		is_refrigerated = frappe.db.get_value("Item", item.item_code, "custom_is_refrigerated_")
		if item_group in ["Narcotics", "Semi Control"] or is_refrigerated:
			frappe.throw(f"Return not allowed for NHRA item {item.item_name}.")

		if days_diff > 14:
			frappe.throw("Items can only be returned before 14 days.")


def validate_dispensing_lots(doc, method):
	# Lot consumption belongs on Delivery Note / stock-updating invoices only.
	if not cint(doc.get("update_stock")):
		return
	validate_sales_invoice_dispensing_lots(doc)


def update_dispensing_lots_on_submit(doc, method):
	if not cint(doc.get("update_stock")):
		return
	process_sales_invoice_dispensing_lots(doc, is_return=bool(doc.get("is_return")))


def restore_dispensing_lots_on_cancel(doc, method):
	if not cint(doc.get("update_stock")):
		return
	reverse_sales_invoice_dispensing_lots(doc)

File: beveren_health/public/js/sales_invoice.js
// Copyright (c) 2026, Beveren Software and contributors
// For license information, please see license.txt

function auto_fill_return_lots(frm) {
	if (!frm.doc.return_against || !frm.doc.items?.length) {
		return;
	}

	frappe.call({
		method: "beveren_health.beveren_health.customize.dispensing_lot.resolve_return_lots_for_lines",
		args: {
			doctype: frm.doc.doctype,
			return_against: frm.doc.return_against,
		},
		callback(r) {
			if (!r || !r.message) {
				return;
			}
			const lot_map = r.message;
			frm.doc.items.forEach((row) => {
				if (row.custom_dispensing_lot) {
					return;
				}
				const ref = row.sales_invoice_item || row.dn_detail;
				if (ref && lot_map[ref]) {
					frappe.model.set_value(
						row.doctype,
						row.name,
						"custom_dispensing_lot",
						lot_map[ref]
					);
				}
			});
			frm.refresh_field("items");
		},
	});
}

frappe.ui.form.on("Sales Invoice", {
	refresh(frm) {
		if (frm.doc.is_return) {
			auto_fill_return_lots(frm);
		}

		frm.set_query("custom_dispensing_lot", "items", (doc, cdt, cdn) => {
			const row = locals[cdt][cdn];
			if (!row || !row.item_code) {
				return { filters: { name: ["in", []] } };
			}

			const filters = {
				item: row.item_code,
			};

			if (frm.doc.is_return) {
				// A return can put stock back onto a delivered lot (full pack sold)
				// or a partially sold lot — include all non-inactive lots.
				filters.status = ["in", ["Active", "Partially Sold", "Delivered"]];
			} else {
				filters.status = ["in", ["Active", "Partially Sold"]];
				filters.remaining_qty = [">", 0];
			}

			if (row.batch_no) {
				filters.batch_no = row.batch_no;
			}

			return { filters };
		});
	},

	update_stock(frm) {
		(frm.doc.items || []).forEach((row) => {
			set_dispensing_lot_reqd(frm, row.doctype, row.name);
		});
	},
});

function set_dispensing_lot_reqd(frm, cdt, cdn) {
	const row = locals[cdt][cdn];
	if (!row || !row.item_code) {
		return;
	}
	// Only require lot when the invoice updates stock (DN already handled stock otherwise).
	if (!frm.doc.update_stock) {
		frm.fields_dict.items.grid.update_docfield_property(
			"custom_dispensing_lot",
			"reqd",
			0,
			cdn
		);
		return;
	}
	frappe.db.get_value("Item", row.item_code, "custom_has_dispense_lot", (r) => {
		const reqd = r.message && r.message.custom_has_dispense_lot ? 1 : 0;
		frm.fields_dict.items.grid.update_docfield_property(
			"custom_dispensing_lot",
			"reqd",
			reqd,
			cdn
		);
	});
}

frappe.ui.form.on("Sales Invoice Item", {
	item_code(frm, cdt, cdn) {
		set_dispensing_lot_reqd(frm, cdt, cdn);
	},

	custom_dispensing_lot(frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		if (!row.custom_dispensing_lot) {
			return;
		}

		frappe.db.get_value(
			"Dispensing Lot",
			row.custom_dispensing_lot,
			["batch_no", "item", "uom", "remaining_qty"],
			(r) => {
				if (!r) {
					return;
				}
				if (r.item && r.item !== row.item_code) {
					frappe.msgprint(
						__("Dispensing Lot belongs to item {0}, not {1}", [r.item, row.item_code])
					);
					frappe.model.set_value(cdt, cdn, "custom_dispensing_lot", "");
					return;
				}
				if (r.batch_no && row.batch_no && r.batch_no !== row.batch_no) {
					frappe.msgprint(
						__("Dispensing Lot batch {0} does not match line batch {1}", [
							r.batch_no,
							row.batch_no,
						])
					);
				}
			}
		);
	},
});
